### PR TITLE
Draft: Testing PR for `display_buffer` approach for display drivers

### PR DESCRIPTION
**Pull request by ayufan**
# What does this implement/fix?

The purpose of this to provide a new way of writing display drivers:

### Design goals

- All existing display should continue working, but their implementation should be marked as deprecated
- Display driver should not manage framebuffer, only send data directly to the target device
- Display driver should use native device rotation where-is possible
- Display driver should declare what is native buffer format for buffer transfer
- Framebuffer should provide an allocated buffer of choice
- Image/Font and all other functions should prefer to use `draw_pixels` or `draw_pixels_format` in all cases which would convert if needed from non-native format to native and render by lines
- Pixel formats should provide a uniform and convenient pixel mapping between all pixel formats
- Transparancy should be implemented using Pixel Format, where the color key should part of Pixel Format implementation, similar to how `PixelA1` and `PixelW1` are implemented

### Performance

- previously took 170ms to render 240x240/RGB565
- framebuffer: with direct blit to framebuffer, it takes 8ms
- framebuffer: with conversion from RGB888 to RGB565, it takes 21ms
- direct to display: with conversion from RGB888 over SPI to RGB565, it takes 56ms
- direct to display: RGB565 to SPI RGB565 it takes 30ms
- displaying framebuffer, it takes around 30ms to be transfered over SPI

- previously took 1700us to render the front from example
- framebuffer: it takes around 400ms now

> The routines for pixel conversion are not yet optimised.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```yaml
display:
  - id: lcd_display
    platform: display_gc9a01
    reset_pin: GPIO13
    cs_pin: GPIO10
    dc_pin: GPIO9
    rotation: 0
    auto_clear_enabled: false

  - id: framebuffer
    platform: display_buffer
    width: 240
    height: 240
    format: RGB565
    auto_clear_enabled: true
    lambda: |-
      start_us = micros();
      it.image(120, 120, id(logo), ImageAlign::CENTER);
      ESP_LOGE("logo_page", "mellow_logo2: %ld", micros() - start_us);

      start_us = micros();
      it.print(120, 120, id(roboto), TextAlign::TOP_CENTER, "Hello World!");
      ESP_LOGE("logo_page", "roboto: %ld", micros() - start_us);

      start_us = micros();
      id(framebuffer).draw(id(lcd_display));
      ESP_LOGE("logo_page", "send: %ld", micros() - start_us);

      ESP_LOGE("logo_page", "refresh: %ld", micros() - refresh_us);
      refresh_us = micros();
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/display_gc9a01/__init__.py
Comment (guillempages) [inline esphome/components/display_gc9a01/__init__.py:1]: IMHO, adding an extra display should be made in a separate PR; this one is big enough as it is ;-)

File: esphome/components/display/__init__.py [hunk 1]
```diff
@@ -45,6 +45,14 @@
     270: display_ns.DISPLAY_ROTATION_270_DEGREES,
 }
 
+PIXEL_TYPES = {
+    "BINARY": display_ns.enum("PixelW1"),
+    "GRAYSCALE": display_ns.enum("PixelW8"),
+    "RGB332": display_ns.enum("PixelRGB332"),
+    "RGB565": display_ns.enum("PixelRGB565_BE"),
```
Comment (guillempages) [inline esphome/components/display/__init__.py:55]: Why "BE" and not the standard RGB565?

File: esphome/components/image/__init__.py [hunk 2]
```diff
@@ -28,14 +28,17 @@
...
+
+ImageType = image_ns.enum("ImageType")
 IMAGE_TYPE = {
     "BINARY": ImageType.IMAGE_TYPE_BINARY,
     "TRANSPARENT_BINARY": ImageType.IMAGE_TYPE_BINARY,
     "GRAYSCALE": ImageType.IMAGE_TYPE_GRAYSCALE,
     "RGB565": ImageType.IMAGE_TYPE_RGB565,
     "RGB24": ImageType.IMAGE_TYPE_RGB24,
     "RGBA": ImageType.IMAGE_TYPE_RGBA,
+    "JPEG": ImageType.IMAGE_TYPE_JPEG,
```
Comment (guillempages) [inline esphome/components/image/__init__.py:41]: I'm not 100% sure about this, but we might want to separate the compression methods into a separate enum/configuration.
(Thinking about my online-image PR here. If it gets separated, it could also implement an "AUTO" type, that would detect the content-type from html).
Also it could separate the input format (JPEG, PNG, GIF,...) from the desired framebuffer format.

For pre-loaded images, both "IMAGE_TYPE" and "IMAGE_FORMAT" could then be made exclusive, so you would still be able to store compressed images in the flash.
And having separate options could also make it easier to instantiate different base classes (e.g. BufferImage vs CompressedImage) later on, based on the specified option.

File: esphome/components/image/image.cpp [hunk 3]
```diff
@@ -124,11 +172,92 @@ Color Image::get_grayscale_pixel_(int x, int y) const {
...
+
+static int jpeg_draw_by_native(JPEGDRAW *pDraw)
+{
+  display::Display *display = (display::Display*)pDraw->pUser;
+  display->draw_pixels_at(
+    pDraw->x, pDraw->y,
+    pDraw->iWidth, pDraw->iHeight,
+    (const uint8_t*)pDraw->pPixels,
+    pDraw->iBpp * pDraw->iWidth * pDraw->iHeight / 8,
+    display->get_native_pixel_format()
```
Comment (guillempages) [inline esphome/components/image/image.cpp]: why do you need to ask "display" for its native_pixel_format, to pass it to its own "draw_pixels_at" method? Couldn't the display method take care of getting its own known parameters?

File: esphome/components/image/__init__.py (hunk 2 above)
Comment (ayufan) [reply esphome/components/image/__init__.py:41]: > I'm not 100% sure about this, but we might want to separate the compression methods into a separate enum/configuration.

I'm actually not sure. It feels just fine for `ImageType`, where `JPEG` is another type of image supported. There's not really an additional `FORMAT` needed, since this is already covered by `PixelFormat`. Image Type being sometimes RAW (RGB565) or Compressed (JPEG) is likely a good place to have this.

As for the image detection (of compressed formats), there could be a method to detect image type based on a few bytes of header, but I don't think that there should be `AUTO` for such.

File: esphome/components/display/__init__.py (hunk 1 above)
Comment (ayufan) [reply esphome/components/display/__init__.py:55]: The current RGB565 implemented in a code is in fact `BE`, where-as most of libraries to default refer `LE`. I did not make my mind yet, but I think I will make `PixelRGB565_LE` to make it explicit. Other pixel formats do have stable endiness.

Correctness: Sometimes it is BE, sometimes it is LE.

File: esphome/components/image/image.cpp (hunk 3 above)
Comment (ayufan) [reply esphome/components/image/image.cpp]: The `draw_pixels_at` is meant to be able to render data buffer of any format. In this case we have a buffer in a native format, thus why the native format is fetched.

The `draw_pixels_at` depending on a format, and a type of display might: do direct draw, do direct bitblt to buffer, or convert to draw, convert to buffer.

Comment (guillempages) [reply esphome/components/image/image.cpp]: But then, you could have an overloaded version of the method, that implicitly uses the display's own native pixel format. It really doesn't feel good, to have to ask a value from a class, only to pass it to a method of that class ;-)

Comment (ayufan) [reply esphome/components/image/image.cpp]: This is Draft, not a complete work. So, you are being slightly nitpicky :) Look at wider change in this work. I have this call right now due for the convenience.

File: esphome/components/display_gc9a01/__init__.py
Comment (ayufan) [reply esphome/components/display_gc9a01/__init__.py:1]: Per. Discord:

> all those commits are to be merged in separate PR: https://github.com/esphome/esphome/pull/4996/commits, each represent a single functional non-breaking and compilable change

File: esphome/components/display/__init__.py [hunk 4]
```diff
@@ -45,6 +45,16 @@
     270: display_ns.DISPLAY_ROTATION_270_DEGREES,
 }
 
+PIXEL_TYPES = {
+    "BINARY": display_ns.enum("PixelW1"),
+    "W2": display_ns.enum("PixelW2"),
+    "W4": display_ns.enum("PixelW4"),
```
Comment (nielsnl68) [inline esphome/components/display/__init__.py:52]: Where stands W2 and W4 for? maybe replace them for more understandable terms?

Comment (ayufan) [reply esphome/components/display/__init__.py:52]: The idea is White 2 bits, or White 4 bits. Which in other is Gray 2 / 4 / 8.

**Comment by xdecock** [comment]
This mr would help a lot with ePapers, i've implemented the 2bit grayscale on an epaper and handling both the color and the protocol in waveshare is a nightmare. Getting better separation would probably help.

I've seen a [Dual buffer pull request](https://github.com/esphome/esphome/pull/4833)  also for the il9xxx which might get easier with you work.

**Comment by nielsnl68** [comment]
hi @ayufan could you please commit this changes also to my temporary display repo
at https://github.com/nielsnl68/esphome-display we can use that as testbed to try all different setups and changes etc.

File: esphome/components/display/__init__.py (hunk 4 above)
Comment (clydebarrow) [reply esphome/components/display/__init__.py:52]: So maybe `GRAYSCALE_4BPP` and the like?

File: esphome/components/display/display.h [hunk 5]
```diff
@@ -480,7 +484,15 @@ class Display {
    */
   bool clip(int x, int y);
 
+  virtual PixelFormat get_native_pixel_format() = 0;
+  bool draw_pixels_at(int x, int y, int w, int h, const uint8_t *data, int stride, PixelFormat data_format, Color color_on = COLOR_ON, Color color_off = COLOR_OFF);
```
Comment (clydebarrow) [inline esphome/components/display/display.h:488]: Can you document what the parameters mean exactly? I would assume `stride` is the line width in the source data, but that would be nice to know. And why are there two color values as parameters?

As a general principle I like to see public methods, especially those with long argument lists, have a documentation comment block. We don't currently have good developer docs for ESPHome but having the public API documented will really help.
